File: Solutions/Android/app/assets/content/scripting/lobbyTmp/systems/movementsystem.lua
MovementSystem = System()

MovementSystem.Update = function(self, dt)

	if GameRunning == false then
		return;
	end
	
	local entities = self:GetEntities() 
	for i = 1, #entities do
		local entity = entities[i]

		local position = self:GetComponent(entity, "Position", 0)
		local velocity = self:GetComponent(entity, "Velocity", 0)

		local px, py, pz = position:GetFloat3()
		local vx, vy, vz = velocity:GetFloat3()

		px = px + vx * dt
		py = py + vy * dt
		pz = pz + vz * dt

		position:SetFloat3(px, py, pz)
	end
end

MovementSystem.Initialize = function(self)
	self:SetName("Movement System")
	self:AddComponentTypeToFilter("Position", FilterType.Mandatory)
	self:AddComponentTypeToFilter("Velocity", FilterType.Mandatory)
	
	print("MovementSystem initialized!")
end

MovementSystem.PostInitialize = function(self)

end

MovementSystem.OnEntityAdded = function(self, entityId)

	if GameRunning == false then
		return;
	end

	print("OnEntityAdded (LUA)")
end

MovementSystem.OnEntityRemoved = function(self, entityId)

	if GameRunning == false then
		return;
	end

	print("OnEntityRemoved")
end

File: Source/Game/Systems/MovementSystem.cpp
#include "MovementSystem.h"

MovementSystem::MovementSystem()
{

}
MovementSystem::~MovementSystem()
{

}

void MovementSystem::Initialize()
{
	/*	Rendersystem wants Position and Velocity	*/
	AddComponentTypeToFilter("Position", ECSL::FilterType::Mandatory);
	AddComponentTypeToFilter("Velocity", ECSL::FilterType::Mandatory);


	printf("MovementSystem initialized!\n");
}

void MovementSystem::Update(float _dt)
{
	auto entities = *GetEntities();

	/*	TODO: Some logic to not update matrix every frame	*/
	for (auto entity : entities)
	{
		float*		Position;
		float*		Velocity;

		Position	=	(float*)GetComponent(entity, "Position", 0);
		Velocity	=	(float*)GetComponent(entity, "Velocity", 0);

		Position[0] += Velocity[0] * _dt;
		Position[1] += Velocity[1] * _dt;
		Position[2] += Velocity[2] * _dt;
	}

}

void MovementSystem::OnEntityAdded(unsigned int _entityId)
{

}

void MovementSystem::OnEntityRemoved(unsigned int _entityId)
{

}